Feature/73 get all jokes

HellHunterMax opened the pull request:
# Created The endpoint for getting all jokes.

## How it works.
1. The Controller takes on the call. asks the Service to get it a list of all JokeReplyDto's.
2. The Service then asks the Repo for a List of all Jokes.
3. The repo gets all Jokes from the Database and includes a User (author) per Joke. ( that it gets all users might be changed in future ?) 
This is done by eager loading (telling the Database that you want to get the linked user.). 
4. Repo returns the list to the Service.
5. Service then maps the Jokes to a JokesReplyDto.
6. Service returns the list of JokesReplyDto.
7. Controller Returns the list to te caller.

## Issues I encounterd during this Quest:
1. I had to figure out mapping again.
2. Mapping the User into a Author inside the joke mapper.
3. Loading the User in the Joke from the DB.
4. not returning a whole user with the Joke.

## how i solved them:
1. I looked at @N1c0z code for the Mapper. and looked at [Automapper Documentation](https://docs.automapper.org/en/stable/Getting-started.html)
2. I [found a Stackoverflow post that helped](https://stackoverflow.com/questions/57257077/add-inner-mapper-to-automapper-mapping-profile)
Added this specific code:
```CreateMap<Joke, JokeReplyDto>()
                .ForPath(destination => destination.Author.Id, options =>
                    options.MapFrom(Source => Source.Author.Id))
                .ForPath(destination => destination.Author.UserName, options =>
                    options.MapFrom(source => source.Author.UserName));
```
3. [Different types of loading found here on MS docs](https://docs.microsoft.com/en-us/ef/core/querying/related-data/)
    [and found some explanation here c-shapcorner](https://www.c-sharpcorner.com/UploadFile/dacca2/lazy-and-eager-loading-in-entity-framework/)
4. Created the JokeAuthorDto to not return the whole user when returning a joke

HellHunterMax committed:
Added Joke Skeleton;

Added Controller, Service, Repo and Dto's empty classes and interfaces to be worked on.

HellHunterMax committed:
Created Tag Skeleton;

Added Controller, Service, repo, Model and Dto's as empty classes and interfaces.

HellHunterMax committed:
Created empty Flag Skeleton and models.

Added Controller, Service, Repository, Interfaces, Model and Dto's.

HellHunterMax committed:
Created empty Reason skeleton;

Created Controller, Service, Repository, Interfaces, Model and Dto's
Also added ControllerBase to the controllers;

HellHunterMax committed:
Created empty LikedJoke skeleton.

Added Controller, Service, Repository, Interfaces, Model and Dto's.

HellHunterMax committed:
Created empty SavedJoke Skeleton.

Added Controller, Service, Repository, Interfaces, Model and Dto's.

HellHunterMax committed:
created empty LikedTags skeleton.

Added Controller, Service, Repository, Interfaces, Model and Dto's.

HellHunterMax committed:
Fix: Added Interfaces to classes.

forgot to add the interface to these classes so added them here.

HellHunterMax committed:
Filled in Flag props

HellHunterMax committed:
changed Id from string to int

HellHunterMax committed:
added LikedJoke Props.

HellHunterMax committed:
Added LikedTags props;

HellHunterMax committed:
added IFlaggable.

This way we will be able to flag multiple Models with the same flags.
Need to add it to User, Joke and Tag.

HellHunterMax committed:
Added Reason props;

Changed Reason String to Name, cant have same name as class.

HellHunterMax committed:
added SavedJoke props;

HellHunterMax committed:
Moved IFlaggable into Interfaces folder.

HellHunterMax committed:
Added Tag props.

HellHunterMax committed:
Added Missing props.

HellHunterMax committed:
Created 1 test for JokesService.GetAllJokesAsync.

HellHunterMax committed:
Added JokesReplyDto props

HellHunterMax committed:
Added Interface methods, implemented placeholders in classes.

Added GetAllJokesAsync to IJokesRepository and IJokesService.
implemented the methods in the service and repo.
changed test to get it working.

HellHunterMax committed:
Added Automapper, implemented it into Service and Test;

HellHunterMax committed:
removed skeleton test class.

HellHunterMax committed:
added scopes for Jokes Repo and Service.

HellHunterMax committed:
moved mapper to a class prop.

HellHunterMax committed:
added Controller GetAllJokesAsync method.

Gets the joke Dto's from the Service and returns Ok with the Dto's.

HellHunterMax committed:
added DataContext DI to JokesRepository.

HellHunterMax committed:
Created GetAllJokesAsync in JokesRepository.

HellHunterMax committed:
Added JokeAuthorDto, added it to JokeReplyDto, added it to mapping.

To make the JokeReply return a whole user is no good so we need a Joke Author Dto for that.
Added info inside profile to change the user into a AuthorDto.

HellHunterMax committed:
Added Eager loading to Jokes in JokesRepository.GetAllJokesAsync.

Might be too much info gatherd from the DB but we could change later.

ofu997 commented:
tested this out and it works.

chrisK00 commented on Server/JokesOnYou.Web.Api/Controllers/JokesController.cs:
```diff
@@ -0,0 +1,33 @@
+
+        [AllowAnonymous]
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<JokeReplyDto>>> GetAllJokesAsync()
```
Suggestion: No need to name the method async considering no one will be awaiting it

HellHunterMax replied:
There is no need indeed.
I'm just appending Async to all async methods no matter where to make it a habit for myself.
I'd rather have to much information then forgetting it somewhere. 

So I think I'll keep it but, you can change my mind of course.

chrisK00 commented on Server/JokesOnYou.Web.Api/JokesOnYou.Web.Api.csproj:
```diff
@@ -14,6 +14,8 @@
 
   <ItemGroup>
+    <PackageReference Include="AutoMapper" Version="10.1.1" />
+    <PackageReference Include="AutoMapper.Extensions.Microsoft.DependencyInjection" Version="8.1.1" />
```
we only need this package

HellHunterMax replied:
ok ill remove and test again ;)

chrisK00 commented on Server/JokesOnYou.Web.Api/Profiles/JokesProfile.cs:
```diff
@@ -0,0 +1,22 @@
+        public JokesProfile()
+        {
+            CreateMap<Joke, JokeReplyDto>()
+                .ForPath(destination => destination.Author.Id, options =>
```
Theres no need to do forpath/config for every property. If we would have to do it for every entity/prop automapper becomes kinda useless. The way automapper works is it compares the property names in both entities, if they match it maps the one to the other.

HellHunterMax replied:
ill try without. i thought i did try it without and didn't work but ill try again!

HellHunterMax replied:
Added 
```
CreateMap<User, JokeAuthorDto>();
```
Works now.

chrisK00 commented on Server/JokesOnYou.Web.Api/Services/JokesService.cs:
```diff
@@ -0,0 +1,32 @@
+        public async Task<IEnumerable<JokeReplyDto>> GetAllJokesAsync()
+        {
+            var jokes = await _jokesRepo.GetAllJokesAsync();
+
```
Probably the best way to handle this is to use automapper https://docs.automapper.org/en/stable/Queryable-Extensions.html. This way we only SELECT the properties we need from the db and we dont even need to manually include navigation props.

HellHunterMax replied:
oh cool thanks! i check this out tonight!

HellHunterMax replied:
Very nice this works. exactly what I was worried about ;)
I tried different as well but that didn't work.
I tried 
```
var jokes = await _context.Jokes.Include( joke => joke.Author.Id)
                                .Include(joke => joke.Author.UserName)
                                .ToListAsync();
```
Must be a way to get this to work? but maybe not.

chrisK00 commented on Server/JokesOnYou.Web.Api/Services/JokesService.cs:
```diff
@@ -0,0 +1,32 @@
+            _mapper = mapper;
+        }
+
+        public async Task<IEnumerable<JokeReplyDto>> GetAllJokesAsync()
```
Suggestion: GetAllJokesDtosAsync

chrisK00 commented on Server/JokesOnYou.Web.Api/Startup.cs:
```diff
@@ -36,13 +37,17 @@ public void ConfigureServices(IServiceCollection services)
+            services.AddScoped<IJokesService, JokesService>();
 
             services.AddMvc()
             .AddJsonOptions(o => {
```
not sure when i missed this, could you explain? (if you wrote it)

HellHunterMax replied:
woeps that should have been commented out in a fix earlier back hmm ill remove it again.

chrisK00 commented on Tests/Server/JokesServiceTests.cs:
```diff
@@ -0,0 +1,77 @@
+        public void GetAllUsers_Returns_ListOfJokesDto()
+        {
+            var repo = new Mock<IJokesRepository>();
+            var service = new JokesService(repo.Object, _mapper);
```
Use the constructor as a setup method, it will drastically reduce the amount of code in our tests

chrisK00 commented on Tests/Server/JokesServiceTests.cs:
```diff
@@ -0,0 +1,77 @@
+                expected.Add(new JokeReplyDto() { Id = joke.Id, Premise = joke.Premise, Punchline = joke.Punchline, Likes = joke.Likes, Dislikes = joke.Dislikes });
+            }
+
+            var actualList = service.GetAllJokesAsync().Result;
```
How about awaiting this async method instead of trying to get the result out?

HellHunterMax replied:
https://stackoverflow.com/questions/29809302/is-there-a-difference-between-calling-result-or-await-on-known-completed-tasks#:~:text=The%20only%20functional%20difference%20that,will%20just%20raise%20the%20exception.

So i think not much difference but await needs the whole test to be async. but ill test it out anyway would like to see.

chrisK00 replied:
> https://stackoverflow.com/questions/29809302/is-there-a-difference-between-calling-result-or-await-on-known-completed-tasks#:~:text=The%20only%20functional%20difference%20that,will%20just%20raise%20the%20exception.
> 
> So i think not much difference but await needs the whole test to be async. but ill test it out anyway would like to see.

actually theres more to it, thats why async/await exists. https://stackoverflow.com/questions/58312479/is-accessing-async-result-considered-as-bad-practice

I think theres a ReturnsAsync() method aswell for Moq so you should be able to figure it out np

HellHunterMax replied:
So the Service doesn't do anything anymore. what should be tested is the mapper now ;)
removed the test.

chrisK00 reviewed (changes requested):
Good job. I like that you dont instantly return all items in the controller, making it easier to debug.

HellHunterMax commented:
Thanks for all the comments ill fix this tonight!

HellHunterMax committed:
Removed specific mapping, added User -> JokeAuthorDto map.

HellHunterMax committed:
JokesRepo does not return whole user anymore.

Added GetAllJokeDtosAsync to service and Repository.
added Interfaces for them.

HellHunterMax committed:
removed the loopfix in startup.

As there is no loop anymore.

HellHunterMax committed:
removed test as GetAllJokes in Service does not do anything.

Should test mapper though.

HellHunterMax commented on Server/JokesOnYou.Web.Api/Repositories/JokesRepository.cs:
```diff
@@ -0,0 +1,43 @@
+            var configuration = new MapperConfiguration(cfg =>
+            {
+                cfg.CreateMap<Joke, JokeReplyDto>();
+                cfg.CreateMap<User, JokeAuthorDto>();
```
Any other way of doing this since i already have the mapping profile?

chrisK00 replied:
Yes. So since we have the profile configured, and automapper is added to DI serviceprovider. We can just inject the mapper using IMapper interface and use it all over the place.

HellHunterMax replied:
Very nice! 
_mapper.ConfigurationProvider worked here I was thinking there should be a way :D

chrisK00 reviewed (approved):
Inject automapper using IMapper and then this is ready for merge!

HellHunterMax committed:
repo now uses DI for mapperconfiguration.
